compiler/src/dmd/constfold.d

```diff
@@ -133,89 +133,129 @@ UnionExp Add(const ref Loc loc, Type type, Expression e1, Expression e2)
     static if (LOG)
     {
         printf("Add(e1 = %s, e2 = %s)\n", e1.toChars(), e2.toChars());
+        printf("type %s t1 %s t2 %s\n", type.toChars(), e1.type.toChars(), e2.type.toChars());
     }
-    if (type.isreal())
-    {
-        emplaceExp!(RealExp)(&ue, loc, e1.toReal() + e2.toReal(), type);
-    }
-    else if (type.isimaginary())
-    {
-        emplaceExp!(RealExp)(&ue, loc, e1.toImaginary() + e2.toImaginary(), type);
-    }
-    else if (type.iscomplex())
+
+    float e1ref() { return cast(float)e1.toReal(); }
+    float e1imf() { return cast(float)e1.toImaginary(); }
+    float e2ref() { return cast(float)e2.toReal(); }
+    float e2imf() { return cast(float)e2.toImaginary(); }
+
+    double e1red() { return cast(double)e1.toReal(); }
```

Comment:
I'm pretty sure I've said this countless times before - and there might even be a PR still open where something similar is being done.

No native floating point arithmetic in the front-end. Not even floating point literals are to be found anywhere.

Reply:
How do you suggest doing a correctly rounded floating point operation?

Reply:
Basically I need addf and addd functions that meet your requirements, and later mul, div and mod.

Reply:
> How do you suggest doing a correctly rounded floating point operation?

We have `dmd.root.longdouble` as a software-emulated floating point type configured to do simple arithmetic at the target's highest precision (the 80-bit floating point comment at the top of that file only applies to dmd), and `real_t` is an alias to that struct.  Then there is `dmd.root.ctfloat` for doing more operations with that type.

Currently the implementation of `toPrec` is wrapped up amongst the built-ins.  It might be better to put that in CTFloat to have a `.truncate()` or `.convert()` method to apply to the internal representation during certain stages of evaluation.  Or like IntegerExp call `normalize()` in the constructor of every new RealExp node - except forward that to `CTFloat.normalize(mode)`.

Do bear in mind that C99/11 does allow for excess precision, and going the other way does cause bugs too (the infamous https://gcc.gnu.org/bugzilla/show_bug.cgi?id=323).  Any fix needs to communicate with the target back-end about what is the correct behaviour (FLT_EVAL_METHOD), i.e: ARM32/64 and RISC-V evaluate to the range and precision of float, SystemZ evaluates to the range and precision of double, x86_64 also evaluates to the range and precision of float, but non-OSX X86 evaluates to the precision of real.

Reply:
Longdouble doesn't actually do software emulation

Reply:
My branch `floatHacking` contains a file with some x87 emulation routines which *do* emulate 80 bit floats in RNE mode

Reply:
> Longdouble doesn't actually do software emulation

[Oh yes it does](https://github.com/gcc-mirror/gcc/blob/531ca06c007d4c4d156637083dcad7f25ac8713d/gcc/d/dmd/root/longdouble.d#L123-L126).

Reply:
A good step to killing some of these warts is to make dmd use longdouble. Currently it can't actually compile because it assumes it's a native D real.

Reply:
I'll save dealing with FLT_EVAL_METHOD variations for the future. Right now, I'm only going to address this particular issue.

> non-OSX X86 evaluates to the precision of real

On my Ubuntu machine it evaluates to the precision of float. This PR produces the same result as gcc with its default settings.

BTW, I could build a float emulator and a double emulator pretty easily, as I already have each in x86 asm. It would just have to be converted to D. An 80 bit emulator is harder, and @maxhaton is working on that.

Reply:
> On my Ubuntu machine it evaluates to the precision of float. This PR produces the same result as gcc with its default settings.
> 

X86-32 rounds to 80-bit real, not X86_64.  This is also what the [docs](https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/x86-Options.html#index-mfpmath-1) seem to suggest, and the test in the bugzilla issue confirms it too.

X86-64 (gcc -m64)
```
-1.000000
-1.000000
0.000000
0.000000
0.000000
0.000000
```

X86-32  (gcc -m32)
```
-1.000000
-1.000000
-1.000000
0.000000
-1.000000
0.000000
```

Reply:
It is strange that D deviates from X86-32 though.

Both -m32 and -m64.
```
-1.000000
-1.000000
0.000000
-1.000000
0.000000
-1.000000
```

Reply:
@WalterBright I've hacked my dmd to use my float routines but it requires a pretty big diff to the backend which I think might result in subtle bugs.

As I said above I think making dmd.backend always use longdouble_soft for float (even though the current impl is actually just still using x87 underneath) would massively reduce the pain of going emulaged.

Reply:
Why would it be a big diff? Just call a function instead of using `+`.

Reply:
@ibuclaw I don't really know what to do with these results.

Reply:
@WalterBright All the compiler currently does go through an alias `real_t` and so on (or targ_longdouble etc) BUT dmd currently cannot compile if this is aliased to anything other than a builtin `real` (any soft type currently can't be used for all targets) - partly because of some things which are assuming (e.g. printf formatting) that the "real" type is native but also because the current "soft" implementation uses x87 underneath which means that the FPU has to be initialized properly. All the bits to fix this are there but currently under a ratsnest of version(microsoftlibc) and so on

also by large I mean maybe 100 lines. Not properly big I suppose but don't want to change behaviour.